gadget/layout.go

```diff
@@ -314,9 +314,15 @@ func resolveVolumeContent(gadgetRootDir, kernelRootDir string, kernelInfo *kerne
 // content and populates VolumeContent.resolvedSource with absolute
 // paths.
 func resolveContentPathOrRef(gadgetRootDir, kernelRootDir string, kernelInfo *kernel.Info, pathOrRef string) (string, error) {
-	// TODO: what to here? we could error instead?
-	if pathOrRef == "" {
-		return "", nil
+
+	// TODO: add kernelRootDir == "" error too once all the higher
+	//       layers in devicestate call gadget.Update() with a
+	//       kernel dir set
+	switch {
+	case gadgetRootDir == "":
+		return "", fmt.Errorf("internal error: gadget root dir cannot beempty")
+	case pathOrRef == "":
+		return "", fmt.Errorf("cannot use empty source")
```

Comment:
didn't we discuss to do this earlier as well,  or is that already done?

Reply:
We did, I can move it into the caller:
```
diff --git a/gadget/layout.go b/gadget/layout.go
index 63aeb85937..c09a391c7a 100644
--- a/gadget/layout.go
+++ b/gadget/layout.go
@@ -297,6 +297,10 @@ func resolveVolumeContent(gadgetRootDir, kernelRootDir string, kernelInfo *kerne
 
        content := make([]ResolvedContent, len(ps.Content))
        for idx := range ps.Content {
+               if ps.Content[idx].UnresolvedSource == "" {
+                       return nil, fmt.Errorf("cannot use empty source in structure %v at index %v", ps, idx)
+               }
+
                resolvedSource, err := resolveContentPathOrRef(gadgetRootDir, kernelRootDir, kernelInfo, ps.Content[idx].UnresolvedSource)
                if err != nil {
                        return nil, fmt.Errorf("cannot resolve content for structure %v at index %v: %v", ps, idx, err)
@@ -318,11 +322,8 @@ func resolveContentPathOrRef(gadgetRootDir, kernelRootDir string, kernelInfo *ke
        // TODO: add kernelRootDir == "" error too once all the higher
        //       layers in devicestate call gadget.Update() with a
        //       kernel dir set
-       switch {
-       case gadgetRootDir == "":
+       if gadgetRootDir == "" {
                return "", fmt.Errorf("internal error: gadget root dir cannot beempty")
-       case pathOrRef == "":
-               return "", fmt.Errorf("cannot use empty source")
        }
 
        // content may refer to "$kernel:<name>/<content>"
```
This duplicates a little bit of the error string below. Happy to change if you prefer it this way.

Reply:
sorry, what I meant is something different, does Validate fails if source or target are empty? or is there a use case for that?

Reply:
I should but I will add a test

Reply:
I pushed the updated test